packages/dds/map/src/directory.ts

```diff
@@ -1396,6 +1577,79 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
+            throw new Error("Rollback op does not match last pending");
+        }
+        if (pendingMessageIds.length === 0) {
+            map.delete(key);
+        }
+    }
+
+    /**
+     * Rollback a local op
+     * @param op - The operation to rollback
+     * @param localOpMetadata - The local metadata associated with the op.
+     */
+     public rollback(op: any, localOpMetadata: unknown) {
+        if (!isDirectoryLocalOpMetadata(localOpMetadata)) {
+            throw new Error("Invalid localOpMetadata");
+        }
+
+        if (op.type === "clear" && localOpMetadata.type === "clear") {
+            if (localOpMetadata.previousStorage === "unknown") {
+                throw new Error("Cannot rollback without previous contents");
+            }
+            localOpMetadata.previousStorage.forEach((localValue, key) => {
+                this.setCore(key, localValue, true);
+            });
+
+            const lastPendingClearId = this.pendingClearMessageIds.pop();
+            if (lastPendingClearId === undefined || lastPendingClearId !== localOpMetadata.pendingMessageId) {
+                throw new Error("Rollback op does match last clear");
+            }
+        } else if ((op.type === "delete" || op.type === "set") && localOpMetadata.type === "edit") {
+            if (localOpMetadata.previousValue === "unknown") {
+                throw new Error("Cannot rollback without previous contents");
+            }
+            if (localOpMetadata.previousValue === undefined) {
+                this.deleteCore(op.key, true);
+            } else {
+                this.setCore(op.key, localOpMetadata.previousValue, true);
+            }
+
+            this.rollbackPendingMessageId(this.pendingKeys, op.key, localOpMetadata.pendingMessageId);
+        } else if (op.type === "createSubDirectory" && localOpMetadata.type === "createSubDir") {
+            if (localOpMetadata.previouslyExisted === "unknown") {
+                throw new Error("Cannot rollback without previous contents");
+            } else if (!localOpMetadata.previouslyExisted) {
+                this.deleteSubDirectoryCore(op.subdirName, true);
+            }
+
+            this.rollbackPendingMessageId(this.pendingSubDirectories, op.subdirName, localOpMetadata.pendingMessageId);
+        } else if (op.type === "deleteSubDirectory" && localOpMetadata.type === "deleteSubDir") {
+            if (localOpMetadata.subDirectory === "unknown") {
+                throw new Error("Cannot rollback without previous contents");
+            } else if (localOpMetadata.subDirectory !== undefined) {
+                this.undeleteSubDirectoryTree(localOpMetadata.subDirectory);
+                // don't need to register events because deleting never unregistered
+                this._subdirectories.set(op.subdirName, localOpMetadata.subDirectory);
+                this.emit("subDirectoryCreated", op.subdirName, true, this);
```

Comment:
The subDirectoryCreated event has been used to indicate a brand new subdirectory with no content. This is different though. Maybe there should be some different event, or there should be creation events on everything within this subdirectory?